### PR TITLE
feat: add ostream operator for Value

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/1336)

### CODE REVIEW

**Comment by scotthart** [comment]
fixes #1303

File: google/cloud/spanner/value.cc [hunk 1]
```diff
@@ -88,6 +88,79 @@ bool Equal(google::spanner::v1::Type const& pt1,
...
+  }
+  return os << std::get<2>(delimiters_type_function(i));
+}
+
+std::ostream& StreamHelper(std::ostream& os, google::protobuf::Value const& v,
+                           google::spanner::v1::Type const& t) {
+  if (v.kind_case() == google::protobuf::Value::kNullValue &&
+      t.code() != google::spanner::v1::ARRAY &&
+      t.code() != google::spanner::v1::STRUCT) {
+    return os;
```
Comment (devbww) [inline google/cloud/spanner/value.cc]: Q. Should this render "NULL"?

File: google/cloud/spanner/value_test.cc [hunk 2]
```diff
@@ -904,6 +912,75 @@ TEST(Value, CommitTimestamp) {
...
+        std::make_tuple("NaNFloat64", Value(std::nan("NaN")), "NaN"),
+        std::make_tuple("String", Value("Seatac Astronomy"),
+                        "Seatac Astronomy"),
+        std::make_tuple("Bytes", Value("DEADBEEF"), "DEADBEEF"),
+        std::make_tuple(
+            "Timestamp",
+            Value(MakeTimestamp(MakeTimePoint(1561147549LL, 0)).value()),
+            "2019-06-21T20:05:49Z"),
+        std::make_tuple("Date", Value(Date(1970, 1, 1)), "1970-01-01")),
+
```
Comment (devbww) [inline google/cloud/spanner/value_test.cc]: Nit: Delete blank line to match other instantiations (or vice versa)?

File: google/cloud/spanner/value.cc (hunk 1 above)
Comment (scotthart) [reply google/cloud/spanner/value.cc]: Perhaps, but for a string field how would I differentiate between a null value and a literal of "NULL"?

File: google/cloud/spanner/value_test.cc (hunk 2 above)
Comment (scotthart) [reply google/cloud/spanner/value_test.cc]: Done.

File: google/cloud/spanner/value.cc (hunk 1 above)
Comment (devbww) [reply google/cloud/spanner/value.cc]: Perhaps a string field renders with '"' delimiters (just like structs and arrays have delimiters)?

File: google/cloud/spanner/value_test.cc [hunk 3]
```diff
@@ -945,13 +945,12 @@ INSTANTIATE_TEST_SUITE_P(
         std::make_tuple("NaNFloat64", Value(std::nan("NaN")), "NaN"),
         std::make_tuple("String", Value("Seatac Astronomy"),
                         "Seatac Astronomy"),
-        std::make_tuple("Bytes", Value("DEADBEEF"), "DEADBEEF"),
+        std::make_tuple("Bytes", Value(Bytes("DEADBEEF")), "DEADBEEF"),
```
Comment (devbww) [inline google/cloud/spanner/value_test.cc]: Perhaps BYTES should render with the `b` prefix (and the quotes) ... `b"DEADBEEF"`.

File: google/cloud/spanner/value.cc [hunk 4]
```diff
@@ -88,6 +88,79 @@ bool Equal(google::spanner::v1::Type const& pt1,
...
+    case google::spanner::v1::FLOAT64:
+      if (v.kind_case() == google::protobuf::Value::kStringValue) {
+        return os << v.string_value();
+      }
+      return os << std::to_string(v.number_value());
+
+    case google::spanner::v1::TIMESTAMP:
+    case google::spanner::v1::DATE:
+    case google::spanner::v1::STRING:
+    case google::spanner::v1::BYTES:
```
Comment (devbww) [inline google/cloud/spanner/value.cc]: We might want to consider escape sequence for non-printable chars.

File: google/cloud/spanner/value_test.cc [hunk 5]
```diff
@@ -944,8 +944,8 @@ INSTANTIATE_TEST_SUITE_P(
                         "-Infinity"),
         std::make_tuple("NaNFloat64", Value(std::nan("NaN")), "NaN"),
         std::make_tuple("String", Value("Seatac Astronomy"),
-                        "Seatac Astronomy"),
-        std::make_tuple("Bytes", Value(Bytes("DEADBEEF")), "DEADBEEF"),
+                        "\"Seatac Astronomy\""),
+        std::make_tuple("Bytes", Value(Bytes("DEADBEEF")), "B\"DEADBEEF"),
```
Comment (devbww) [inline google/cloud/spanner/value_test.cc]: It looks like there should be a trailing `\"` there.

File: google/cloud/spanner/value.cc [hunk 6]
```diff
@@ -88,6 +88,83 @@ bool Equal(google::spanner::v1::Type const& pt1,
   }
 }
 
+std::ostream& StreamHelper(std::ostream& os, google::protobuf::Value const& v,
+                           google::spanner::v1::Type const& t) {
+  if (v.kind_case() == google::protobuf::Value::kNullValue &&
+      t.code() != google::spanner::v1::ARRAY &&
+      t.code() != google::spanner::v1::STRUCT) {
+    return os << "";
```
Comment (devbww) [inline google/cloud/spanner/value.cc]: Why stream the `""`?  Did you mean `"NULL"`?

File: google/cloud/spanner/value.cc [hunk 7]
```diff
@@ -88,6 +88,83 @@ bool Equal(google::spanner::v1::Type const& pt1,
...
+  }
+
+  int i = 0;
```
Comment (devbww) [inline google/cloud/spanner/value.cc]: I think the observation was that the `i` could be local to the `for` loops.  Its use after the loop could be replaced with another expression given that the loops don't `break`.  But it also looks the the "size - 1" loops could be replaced with full loops by using the `std::string sep` idiom.

File: google/cloud/spanner/value.cc [hunk 8]
```diff
@@ -88,6 +88,83 @@ bool Equal(google::spanner::v1::Type const& pt1,
...
+    case google::spanner::v1::INT64:
+      return os << v.string_value();
+
+    case google::spanner::v1::FLOAT64:
+      if (v.kind_case() == google::protobuf::Value::kStringValue) {
+        return os << v.string_value();
+      }
+      return os << std::to_string(v.number_value());
+
+    case google::spanner::v1::TIMESTAMP:
```
Comment (devbww) [inline google/cloud/spanner/value.cc]: Might as well group these with `INT64`.

File: google/cloud/spanner/value.h [hunk 9]
```diff
@@ -314,6 +314,18 @@ class Value {
    */
   friend void PrintTo(Value const& v, std::ostream* os);
 
+  /**
+   * Outputs string representation of a given Value to the provided stream.
```
Comment (devjgm) [inline google/cloud/spanner/value.h]: I think this should copy the warning from the `PrintTo` function:

```
@warning This is intended for debugging and human consumption only, not
    machine consumption as the output format may change without notice.
```

**Comment by scotthart** [comment]
created #1345 to track opportunity to improve bytes formatting.

File: google/cloud/spanner/value.h (hunk 9 above)
Comment (scotthart) [reply google/cloud/spanner/value.h]: Done.

File: google/cloud/spanner/value.cc (hunk 7 above)
Comment (devjgm) [reply google/cloud/spanner/value.cc]: Right. For example, I think the array case might look like this:

```cc
    case google::spanner::v1::ARRAY: {
      os << '[';
      char const* sep = nullptr;
      for (auto const& val : v.list_value().values()) {
        if (sep != nullptr) os << sep;
        StreamHelper(os, val, t.array_element_type());
        sep = ", ";
      }
      return os << ']';
    }
```

Struct could simplify to something like this as well.

File: google/cloud/spanner/value.cc (hunk 6 above)
Comment (scotthart) [reply google/cloud/spanner/value.cc]: Done.

File: google/cloud/spanner/value.cc (hunk 7 above)
Comment (scotthart) [reply google/cloud/spanner/value.cc]: Done.

File: google/cloud/spanner/value.cc (hunk 8 above)
Comment (scotthart) [reply google/cloud/spanner/value.cc]: Done.

File: google/cloud/spanner/value_test.cc (hunk 5 above)
Comment (scotthart) [reply google/cloud/spanner/value_test.cc]: Done.

File: google/cloud/spanner/value.cc [hunk 10]
```diff
@@ -88,6 +88,75 @@ bool Equal(google::spanner::v1::Type const& pt1,
...
+        StreamHelper(os, e, t.array_element_type());
+        delimiter = ", ";
+      }
+      return os << ']';
+    }
+    case google::spanner::v1::STRUCT: {
+      const char* delimiter = "";
+      os << '(';
+      for (int i = 0;
+           v.list_value().values_size() > 0 &&
```
Comment (devbww) [inline google/cloud/spanner/value.cc]: This looks redundant.

File: google/cloud/spanner/value.cc [hunk 11]
```diff
@@ -88,6 +88,77 @@ bool Equal(google::spanner::v1::Type const& pt1,
...
+    case google::spanner::v1::TIMESTAMP:
+    case google::spanner::v1::DATE:
+    case google::spanner::v1::INT64:
+      return os << v.string_value();
+
+    case google::spanner::v1::FLOAT64:
+      if (v.kind_case() == google::protobuf::Value::kStringValue) {
+        return os << v.string_value();
+      }
+      return os << std::to_string(v.number_value());
```
Comment (devjgm) [inline google/cloud/spanner/value.cc]: Is this `std::to_string` needed?

**Comment by devjgm** [comment]
ftr, I didn't mean `Thanks, Scott?`, I meant `Thanks, Scott!`

File: google/cloud/spanner/value.cc (hunk 11 above)
Comment (scotthart) [reply google/cloud/spanner/value.cc]: Depends. 
Without the to_string 42.0 is printed as 42 and 5.2 as 5.2.
With to_string you get 42.000000 and 5.200000

Comment (devjgm) [reply google/cloud/spanner/value.cc]: I'd go w/ the first output, w/o the std::to_string, but I'll let you decide.

File: google/cloud/spanner/value.cc (hunk 10 above)
Comment (scotthart) [reply google/cloud/spanner/value.cc]: Done.

File: google/cloud/spanner/value.cc (hunk 11 above)
Comment (scotthart) [reply google/cloud/spanner/value.cc]: Done.

File: google/cloud/spanner/value_test.cc [hunk 12]
```diff
@@ -904,6 +912,81 @@ TEST(Value, CommitTimestamp) {
...
+        std::make_tuple("InfinityFloat64",
+                        Value(std::numeric_limits<double>::infinity()),
+                        "Infinity"),
+        std::make_tuple("NegativeInfinityFloat64",
+                        Value(-std::numeric_limits<double>::infinity()),
+                        "-Infinity"),
+        std::make_tuple("NaNFloat64", Value(std::nan("NaN")), "NaN"),
+        std::make_tuple("String", Value("Seatac Astronomy"),
+                        "\"Seatac Astronomy\""),
+        std::make_tuple("Bytes", Value(Bytes("DEADBEEF")), "B\"DEADBEEF\""),
```
Comment (devbww) [inline google/cloud/spanner/value_test.cc]: Creating a `Bytes` value from an array `a[N]` is equivalent to `Bytes(&a[0], &a[N])`.  So, `Bytes("DEADBEEF")` would mean something like `Bytes({'D','E','A','D','B','E','E','F','\0'})`.  That is, there is a NUL in there, so the expected value should not match.

Use `Bytes(std::string("DEADBEEF"))` instead, at least until we can determine whether this is a bug or a feature.

File: google/cloud/spanner/value.cc [hunk 13]
```diff
@@ -88,6 +88,73 @@ bool Equal(google::spanner::v1::Type const& pt1,
   }
 }
 
+std::ostream& StreamHelper(std::ostream& os, google::protobuf::Value const& v,
+                           google::spanner::v1::Type const& t) {
+  if (v.kind_case() == google::protobuf::Value::kNullValue &&
+      t.code() != google::spanner::v1::ARRAY &&
```
Comment (devjgm) [inline google/cloud/spanner/value.cc]: I believe there's a difference between an empty array and a null array, and we probably want to print them differently too. I think all NULL values should print the string "NULL", including arrays and structs. And empty values should print as "[]" or "()". To make this work, I think you may just need to remove the two ARRAY and STRUCT conditions here.

WDYT?

File: google/cloud/spanner/value.cc [hunk 14]
```diff
@@ -88,6 +88,71 @@ bool Equal(google::spanner::v1::Type const& pt1,
...
+      return os << v.number_value();
+
+    case google::spanner::v1::STRING:
+      return os << "\"" << v.string_value() << "\"";
+
+    case google::spanner::v1::BYTES:
+      return os
+             << "B\""
+             << internal::BytesFromBase64(v.string_value())->get<std::string>()
+
```
Comment (devjgm) [inline google/cloud/spanner/value.cc]: Remove blank line.

**Pull request merged**